File: services/python-fastapi/app/inference.py
import json
from dataclasses import dataclass
from typing import NamedTuple

import aiohttp
from pydantic import ValidationError

from app.protocol import ErrorKind, ErrorStage, Prediction


@dataclass(slots=True, frozen=True)
class InferenceError(Exception):
    stage: ErrorStage
    kind: ErrorKind
    message: str
    inference_status: int | None
    retryable: bool

    def __post_init__(self) -> None:
        super().__init__(self.message)


class InferenceErrorDetails(NamedTuple):
    stage: ErrorStage
    kind: ErrorKind
    inference_status: int | None
    retryable: bool


def _classify_client_error(exc: Exception) -> InferenceErrorDetails:
    if isinstance(exc, TimeoutError):
        return InferenceErrorDetails(
            stage=ErrorStage.INFERENCE_REQUEST,
            kind=ErrorKind.TIMEOUT,
            inference_status=None,
            retryable=True,
        )
    if isinstance(exc, aiohttp.ClientResponseError):
        status = exc.status
        if status == 429:
            return InferenceErrorDetails(
                stage=ErrorStage.INFERENCE_REQUEST,
                kind=ErrorKind.HTTP_429,
                inference_status=status,
                retryable=True,
            )
        if 500 <= status < 600:
            return InferenceErrorDetails(
                stage=ErrorStage.INFERENCE_REQUEST,
                kind=ErrorKind.HTTP_5XX,
                inference_status=status,
                retryable=True,
            )
        return InferenceErrorDetails(
            stage=ErrorStage.INFERENCE_REQUEST,
            kind=ErrorKind.HTTP_5XX,
            inference_status=status,
            retryable=False,
        )
    if isinstance(exc, aiohttp.ClientError):
        return InferenceErrorDetails(
            stage=ErrorStage.INFERENCE_REQUEST,
            kind=ErrorKind.CONNECTION_RESET,
            inference_status=None,
            retryable=True,
        )
    if isinstance(exc, json.JSONDecodeError | ValidationError):
        return InferenceErrorDetails(
            stage=ErrorStage.INFERENCE_RESPONSE_PARSE,
            kind=ErrorKind.PARSE_ERROR,
            inference_status=None,
            retryable=False,
        )
    return InferenceErrorDetails(
        stage=ErrorStage.INFERENCE_REQUEST,
        kind=ErrorKind.CONNECTION_RESET,
        inference_status=None,
        retryable=True,
    )


def _classify_status(status: int) -> InferenceErrorDetails | None:
    if 200 <= status < 300:
        return None
    if status == 429:
        return InferenceErrorDetails(
            stage=ErrorStage.INFERENCE_REQUEST,
            kind=ErrorKind.HTTP_429,
            inference_status=status,
            retryable=True,
        )
    if 500 <= status < 600:
        return InferenceErrorDetails(
            stage=ErrorStage.INFERENCE_REQUEST,
            kind=ErrorKind.HTTP_5XX,
            inference_status=status,
            retryable=True,
        )
    return InferenceErrorDetails(
        stage=ErrorStage.INFERENCE_REQUEST,
        kind=ErrorKind.HTTP_5XX,
        inference_status=status,
        retryable=False,
    )


def _raise_inference_error(details: InferenceErrorDetails, message: str) -> None:
    raise InferenceError(
        stage=details.stage,
        kind=details.kind,
        message=message,
        inference_status=details.inference_status,
        retryable=details.retryable,
    )


async def run_inference(
    inference_client: aiohttp.ClientSession, cpu_passes: int, body: bytes
) -> Prediction:
    try:
        async with inference_client.post(
            "/infer",
            data=body,
            headers={"x-cpu-passes": str(cpu_passes)},
        ) as response:
            if details := _classify_status(response.status):
                _raise_inference_error(details, f"HTTP {response.status}")
            payload = await response.read()
        return Prediction.model_validate_json(payload)
    except InferenceError:
        raise
    except json.JSONDecodeError as exc:
        stage, kind, status, retryable = _classify_client_error(exc)
        raise InferenceError(
            stage=stage,
            kind=kind,
            message=str(exc) or exc.__class__.__name__,
            inference_status=status,
            retryable=retryable,
        ) from exc
    except Exception as exc:
        stage, kind, status, retryable = _classify_client_error(exc)
        raise InferenceError(
            stage=stage,
            kind=kind,
            message=str(exc) or exc.__class__.__name__,
            inference_status=status,
            retryable=retryable,
        ) from exc

File: services/typescript-bun/src/inference.ts
import {
  errorMessage,
  parsePrediction,
  type ErrorKind,
  type ErrorStage,
  type Prediction,
} from "./protocol.ts";

export type InferenceRunner = (
  body: Uint8Array,
  cpuPasses: number,
  signal?: AbortSignal,
) => Promise<Prediction>;

export interface InferenceErrorDetails {
  readonly stage: ErrorStage;
  readonly kind: ErrorKind;
  readonly message: string;
  readonly inferenceStatus: number | null;
  readonly retryable: boolean;
}

export class InferenceError extends Error implements InferenceErrorDetails {
  public readonly stage: ErrorStage;
  public readonly kind: ErrorKind;
  public readonly inferenceStatus: number | null;
  public readonly retryable: boolean;

  public constructor(details: InferenceErrorDetails) {
    super(details.message);
    this.name = "InferenceError";
    this.stage = details.stage;
    this.kind = details.kind;
    this.inferenceStatus = details.inferenceStatus;
    this.retryable = details.retryable;
  }
}

// Bun.fetch uses HTTP/1.1 with keep-alive connection pooling (sized by
// BUN_CONFIG_MAX_HTTP_REQUESTS). This matches the Python gateway's aiohttp
// client; the Rust gateway is the outlier with reqwest's
// http2_prior_knowledge(). Keeping HTTP/1.1 here makes the TS-vs-Python
// comparison apples-to-apples.
export function createInferenceRunner(
  baseUrl: string,
  timeoutMs: number,
  fetchImpl: typeof fetch = fetch,
): InferenceRunner {
  const url = `${baseUrl}/infer`;
  return (body, cpuPasses, signal) =>
    runInference(fetchImpl, url, timeoutMs, body, cpuPasses, signal);
}

export async function runInference(
  fetchImpl: typeof fetch,
  url: string,
  timeoutMs: number,
  body: Uint8Array,
  cpuPasses: number,
  signal?: AbortSignal,
): Promise<Prediction> {
  const timeoutSignal = AbortSignal.timeout(timeoutMs);

  let response: Response;
  try {
    response = await fetchImpl(url, {
      method: "POST",
      headers: { "x-cpu-passes": String(cpuPasses) },
      body,
      signal: signal === undefined ? timeoutSignal : AbortSignal.any([signal, timeoutSignal]),
    });
  } catch (error) {
    throw classifyFetchError(error);
  }

  if (!response.ok) {
    throw new InferenceError({
      stage: "inference_request",
      kind: classifyStatus(response.status),
      message: `inference returned status ${response.status}`,
      inferenceStatus: response.status,
      retryable: response.status === 429 || response.status >= 500,
    });
  }

  try {
    return parsePrediction(await response.json());
  } catch (error) {
    throw new InferenceError({
      stage: "inference_response_parse",
      kind: "parse_error",
      message: errorMessage(error),
      inferenceStatus: null,
      retryable: false,
    });
  }
}

function classifyStatus(status: number): ErrorKind {
  if (status === 429) {
    return "http_429";
  }
  return "http_5xx";
}

function classifyFetchError(error: unknown): InferenceError {
  const message = errorMessage(error);
  if (error instanceof Error && (error.name === "TimeoutError" || error.name === "AbortError")) {
    return new InferenceError({
      stage: "inference_request",
      kind: "timeout",
      message: message || error.name,
      inferenceStatus: null,
      retryable: true,
    });
  }
  return new InferenceError({
    stage: "inference_request",
    kind: "connection_reset",
    message,
    inferenceStatus: null,
    retryable: true,
  });
}
